### ISSUE TITLE
Perturbation prediction tutorial

### ISSUE THREAD

**Issue by kexinhuang12345** [bowang-lab/scGPT#36]
Hi! Thanks for open-sourcing the great work! 

For the perturbation prediction tutorial, I notice that in the training loop:

```
batch_data.to(device)
x: torch.Tensor = batch_data.x  # (batch_size * n_genes, 2)
ori_gene_values = x[:, 0].view(batch_size, n_genes)
pert_flags = x[:, 1].long().view(batch_size, n_genes)
```

it seems like x is a shape of (batch_size * n_genes, 1) from the data loader, is there any processing that you have done? What should pert_flags look like? 

Also, for a conceptual clarifying question, is the perturbation limited to the set of genes in the gene expression readout HVGs? 

Thanks in advance!!

**Comment by subercui** [comment]
Hi @kexinhuang12345 , thanks for your great work GEARS as well, which inspired us a lot! We used the gears data loader here and the x is (size of first dim, 2) here on my side.

![image](https://github.com/bowang-lab/scGPT/assets/11674033/3c77d240-40c6-4bc9-b8e4-de4f3058ee0e)

I am testing using `gears 0.0.1` . Did you update it in later versions?

**Comment by kexinhuang12345** [comment]
I see, thanks for the clarification! Yes, in the old dataloader we are saving the multi-hot pert vector but we figured that we can infer that from the pert variable so we just remove it to save some storage spaces. Thanks again!

**Comment by kexinhuang12345** [comment]
No problem! Yes, since 0.0.4 we updated the loader. Maybe a quick fix would be to use an try and except for the shape of x? In the except, you can derive the multi-hot encoding from the pert_idx variable with https://github.com/snap-stanford/GEARS/blob/c7ca19cbcd6f4da3030d0ebc90b2c2cd0b47a8d8/gears/pertdata.py#L360-L364 

But since the scGPT install requires cell-gears==0.0.1, it should also be fine!

**Comment by subercui** [comment]
Thank you! I'll update this to support the new versions

**Comment by kexinhuang12345** [comment]
Thank you!!

**Comment by kexinhuang12345** [comment]
One additional question: it seems it adds a condition token on the perturbed gene, but what if this perturbed gene is not in the sequenced gene expression profile? does scGPT work for them currently? thanks!

**Comment by subercui** [comment]
Thank you for the question! Currently, we haven't tested that case, but I would imagine it is possible if only one or two perturbed genes are there. The gene tokens and perturbation flag tokens can be readily appended to the input sequence. I think the slightly trick part would be the value of these genes. If there is no sequencing read, a special value or simply zero can be used (They should be fine considering only appearing at one or two positions). Then, the same fine-tuning process can be utilized, i.e. predicting the perturbation readout at those positions that have sequencing reads.

Sorry for the late reply. I hope this makes sense to you and please let me know any further quesstions

**Comment by kexinhuang12345** [comment]
Thanks for the clarification!

**Comment by yanwu2014** [comment]
Thanks for the clarification! So when padding the input sequence for perturbations that aren't measured in the expression data, do we need to include the control group?

**Comment by subercui** [comment]
> Thanks for the clarification! So when padding the input sequence for perturbations that aren't measured in the expression data, do we need to include the control group?

Hi, since the training works with pairs of control and perturbed cells, I think you would need to include the control group
